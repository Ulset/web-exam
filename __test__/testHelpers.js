const userData = {id: 3, firstname: "Sander", lastname: "Ulset", email: "sander.ulset@gmail.com"}
const userToken = "Bearer testTokentihi"
let users = [
    {"id": 1, "firstname": "Testmann", "lastname": "TestPrivatMeldingson", "email": "jeg@finnes.ikke"},
    {"id": 2, "firstname": "Sander", "lastname": "Ulset", "email": "sander.ulset@verisure.no"}
]
let messages = []

const userApi = {
    listUsers: async () => {return users},
    getProfileData: async (userToken) => {return users[0]}
}

const messageApi = {

}


module.exports = {userData, userToken, userApi, messageApi, users}