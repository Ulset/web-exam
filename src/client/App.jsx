import {Link, useHistory} from "react-router-dom";
import {Route, Switch} from "react-router";
import UserList from "./list_uses/ListUsers";
import React from "react";
import NewMessage from "./new_message/NewMessage";
import MyMessages from "./my_messages/MyMessages";

export const App = ({userToken, userData, userApi, messageApi}) => {
    if (!userToken) {
        //If the user isnt logged in, force to the log in page.
        //TODO Add some server authentication if i got the time.
        const history = useHistory()
        history.push("/login")
    }

    if(!userData){
        //If the userdata is still loading, display a message. This prevents components that
        // are dependent on userData to crash on reload.
        return <div>
            <p>Laster brukerdata, vent litt</p>
        </div>
    }

    //Main return statement, this is where the magic starts :P
    return <div>
        <h1>Cool chat app</h1>
        <Link to={'/'}><p>Hjem</p></Link>
        <p>Logget inn som {userData.firstname} {userData.lastname}</p>
        <br/>
        <Switch>
            <Route exact path={'/'}>
                <Link to={'/private_message'}><p>Ny privat melding</p></Link>
                <Link to={'/my_messages'}><p>Chat</p></Link>
                <Link to={'/list_users'}><p>Liste over brukere</p></Link>
            </Route>
            <Route path={'/list_users'}>
                <UserList userApi={userApi} userData={userData}/>
            </Route>
            <Route path={'/private_message'}>
                <NewMessage userApi={userApi} messageApi={messageApi} userData={userData}/>
            </Route>
            <Route path={'/my_messages'}>
                <MyMessages messagesApi={messageApi} userData={userData}/>
            </Route>
        </Switch>
    </div>
}