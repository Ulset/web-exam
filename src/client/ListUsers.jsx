import React, {useState} from 'react'

const ListUsers = ({userApi, userData}) => {
    let [users, setUsers] = useState()
    const selfId = userData.id //The id of the currently logged in user.
    if(!users){
        /*If there are no loaded users, load from server*/
        userApi.listUsers().then(
            d => {
                let usersHtml = d.map(u => {
                    let {id, firstname, lastname, email} = u
                    return <li key={id}><p>{firstname} {lastname}, {email}{selfId===id ? ' (deg)' : ''}</p></li>
                })
                setUsers(usersHtml)
            }
        )
    }
    //Return statement, will tell you if you have no friends
    return <div>
        <h4>Eksisterende brukere</h4>
        <ul>{users ?
            users
            : <p>Laster..</p>}</ul>
        {users && users.length === 1 ? <p>Det er visst bare deg her! Trenger to for Ã¥ chatte! :)</p> : ''}
    </div>
}

export default ListUsers