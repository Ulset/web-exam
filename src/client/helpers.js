const {useEffect} = require("react");
const {useState} = require("react");

const google_ident = {
    discoveryURL: "https://accounts.google.com/.well-known/openid-configuration",
    clientId: "53213695995-hv7g2v86afjaepuacln2go95fjul7hko.apps.googleusercontent.com"
}


module.exports = {google_ident}