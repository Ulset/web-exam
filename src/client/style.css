body {
    background-color: #EFEFEF;
}

.main_app {
    width: 500px;
    background-color: #FFFFFF;
    border-radius: 0.5rem;
    padding: 10px;
    margin: 50px auto auto;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.25);
}

.chat {
    height: 250px;
    width: 100%;
    display: grid;
    grid-template-rows: 1fr 40px;
    grid-template-columns: 1fr auto;
}

.chat > div {
    grid-column-start: 1;
    grid-column-end: 3;
    height: 100%;
    width: 100%;
    overflow-y: auto;
}