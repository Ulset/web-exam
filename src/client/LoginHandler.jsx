import React, {useEffect} from "react";
import {fetchJson} from "./helpers";
import {useHistory} from "react-router";
import {Link} from "react-router-dom";

const google_ident = {
    discoveryURL: "https://accounts.google.com/.well-known/openid-configuration",
    clientId: "53213695995-hv7g2v86afjaepuacln2go95fjul7hko.apps.googleusercontent.com"
}

async function redirect_to_google_login() {
    const {discoveryURL, clientId} = google_ident;
    const {authorization_endpoint} = await fetchJson(discoveryURL)
    const params = {
        client_id: clientId,
        response_type: "token",
        scope: "openid email profile",
        redirect_uri: window.location.origin+"/login",
    };
    window.location.href = authorization_endpoint + "?" + new URLSearchParams(params);
}

const LoginHandler = ({setUserToken}) => {
    if (window.location.href.includes("#")) {
        //If user has already logged inn, parse the supplied userdata
        const hash = Object.fromEntries(new URLSearchParams(window.location.hash.substr(1)));
        const {access_token} = hash
        const access_token_formatted = `Bearer ${access_token}`
        if(hash){
            useEffect(()=>{setUserToken(access_token_formatted)})
            console.log("done")
            const history = useHistory();
            history.push("/")
        }
        return null;
    }

    return <div>
        <h1>Du er desverre ikke logget inn!</h1>
        <button onClick={redirect_to_google_login}>Logg inn med Google</button>
    </div>
}

export default LoginHandler