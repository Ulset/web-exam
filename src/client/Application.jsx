import React, {useEffect, useState} from "react";
import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";
import LoginHandler from "./LoginHandler";
import {App} from "./App";

function Application({userApi, messageApi}) {
    /*The top level of this react app*/

    const [userToken, setUserToken] = useState("");
    const [userData, setUserData] = useState()
    if (!userToken) {
        //If there is no userToken, see if one is saved in localstorage.
        const userToken_stored = localStorage.getItem("userToken")
        if (userToken_stored) {
            //If found, use this as userToken.
            setUserToken(userToken_stored)
        }
    }
    useEffect(() => {
        //If a user token is supplied but no userData is available, ask server who this person is.
        if (!userData && userToken) {
            userApi.getProfileData(userToken).then(data => {
                if (data === "Switch token") {
                    //Too little time to do this better, a Google token only lasts for 1 hour, so need to switch sometimes
                    localStorage.clear()
                    window.location.reload()
                } else {
                    setUserData(data)
                }
            })
        }
        if (userToken) {
            //Every time the userToken is changed (and is something), set this as the userToken in localstorage.
            localStorage.setItem("userToken", userToken)
        }
    }, [userToken])
    //Main return statement. App component will automatically redirect to /login if no userToken is supplied.
    return <div>
        <BrowserRouter>
            <Switch>
                <Route path={'/login'}>
                    <LoginHandler setUserToken={setUserToken} userToken={userToken}/>
                </Route>
                <Route>
                    <App userToken={userToken} userApi={userApi} messageApi={messageApi} userData={userData}/>
                </Route>
            </Switch>
        </BrowserRouter>
    </div>
}

export default Application