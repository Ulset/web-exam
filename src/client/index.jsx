import React, {useEffect, useState} from 'react'
import ReactDom from 'react-dom'
import {Route, Switch, useHistory} from "react-router";
import {BrowserRouter} from "react-router-dom";
import LoginHandler from "./LoginHandler";
import {App} from "./App";

function Index(){
    const [userToken, setUserToken] = useState();

    return <div>
        <BrowserRouter>
            <Switch>
                <Route path={'/login'}>
                    <LoginHandler setUserToken={setUserToken} />
                </Route>
                <Route>
                    <App userToken={userToken}/>
                </Route>
            </Switch>
        </BrowserRouter>
    </div>
}

ReactDom.render(<Index />, document.getElementById("app"))