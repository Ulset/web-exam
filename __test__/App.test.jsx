import {describe, it} from "@jest/globals";
import {act} from "react-dom/test-utils";
import {App} from "../src/client/App";
import React from "react";
import ReactDOM from "react-dom";
import {BrowserRouter} from "react-router-dom";
import {userData, userToken} from "./testHelpers";

describe("Test App component", ()=>{
    it("Shows currently logged in user", async ()=>{
        const container = document.createElement("div")
        document.body.appendChild(container)
        await act(async ()=>{
            ReactDOM.render(<BrowserRouter>
                <App userToken={userToken} userData={userData}/>
            </BrowserRouter>, container)
        })
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div > p").innerHTML).toEqual(`Logget inn som ${userData.firstname} ${userData.lastname}`)
    })

    it("Shows waiting for userdata screen", async ()=>{
        const container = document.createElement("div")
        document.body.appendChild(container)
        await act(async ()=>{
            ReactDOM.render(<BrowserRouter>
                <App userToken={userToken} userData={null}/>
            </BrowserRouter>, container)
        })
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("p").innerHTML).toEqual("Laster brukerdata, vent litt")
    })
})