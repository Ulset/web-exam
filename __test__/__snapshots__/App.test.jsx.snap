// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App component Shows currently logged in user 1`] = `"<div class=\\"main_app\\"><h1>Cool chat app</h1><a href=\\"/\\"><p>Hjem</p></a><p>Logget inn som Sander Ulset</p><button>Logg ut</button><br><br><a href=\\"/private_message\\"><p>Ny privat melding</p></a><a href=\\"/my_messages\\"><p>Chat</p></a><a href=\\"/list_users\\"><p>Liste over brukere</p></a></div>"`;

exports[`Test App component Shows waiting for userdata screen 1`] = `"<div><p>Laster brukerdata, vent litt</p></div>"`;
